### PR TITLE
Allow users to access VariantDebugInfo by setting a flag

### PR DESCRIPTION
**Pull request by e7mac**
PRD: https://leanplum.atlassian.net/wiki/spaces/PRODDEV/pages/548503556/Tinder+Expose+granular+information+on+variables+and+A+B+tests+to+SDK

We allow a flag to be set in "includeVariantDebugInfo" that retrieves content assignment details from the server. We then allow access to this object through Leanplum.variantDebugInfo

### CODE REVIEW

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 1]
```diff
@@ -921,6 +938,16 @@ private static void applyContentInResponse(JSONObject response, boolean alwaysAp
...
     List<Map<String, Object>> variants = JsonConverter.listFromJsonOrDefault(
         response.optJSONArray(Constants.Keys.VARIANTS));
+    VariantDebugInfo variantDebugInfo = null;
+    try {
+      String variantDebugInfoString = response.optString(Constants.Keys.VARIANT_DEBUG_INFO);
+      if (variantDebugInfoString != null) {
+        Gson gson = new Gson();
+        variantDebugInfo = gson.fromJson(variantDebugInfoString, VariantDebugInfo.class);
+      }
+    } catch (Throwable t) {
```
Comment (lan17) [inline AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: Avoid catching throwable, instead catch specific exception that `fromJson` may throw.

File: AndroidSDKCore/build.gradle [hunk 2]
```diff
@@ -30,6 +30,7 @@ dependencies {
     // Compile dependencies will be added as dependency in pom file.
     api "com.android.support:support-v4:[22.0.0,${SUPPORT_LIBRARY_VERSION}]"
     api "com.android.support:appcompat-v7:[22.0.0,${SUPPORT_LIBRARY_VERSION}]"
+    compile 'com.google.code.gson:gson:2.8.5'
```
[the remark this answers is not in the document]
Comment (e7mac) [reply AndroidSDKCore/build.gradle]: This is the reliability vs size tradeoff I mentioned earlier. If we want to parse to models, this is the standard way to do it in Android.

As an aside, Gson is a pretty commonly used library so customers are likely already using it.

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java [hunk 3]
```diff
@@ -921,6 +938,16 @@ private static void applyContentInResponse(JSONObject response, boolean alwaysAp
         response.optJSONObject(Constants.Keys.REGIONS));
     List<Map<String, Object>> variants = JsonConverter.listFromJsonOrDefault(
         response.optJSONArray(Constants.Keys.VARIANTS));
+    VariantDebugInfo variantDebugInfo = null;
+    try {
+      String variantDebugInfoString = response.optString(Constants.Keys.VARIANT_DEBUG_INFO);
+      if (variantDebugInfoString != null) {
+        Gson gson = new Gson();
```
[the remark this answers is not in the document]
Comment (e7mac) [reply AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: I considered that, and decided that the current usage of Gson is rare enough that I would rather not create it for everyone, but instead create on need. However, in the future, we will migrate to a static instance when it is more common.

File: AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java (hunk 1 above)
Comment (e7mac) [reply AndroidSDKCore/src/main/java/com/leanplum/Leanplum.java]: What's the reasoning behind this?
My goal here is to not block the rest of the leanplum response for a debug-info object. I could catch a list of particular exceptions, but my priority is to not block the execution here.

**Pull request closed without merging**